### PR TITLE
`configcli` python library and command line utility.

### PR DESCRIPTION
**Pull request by kcsraju**
## BDvcli 
These bdvcli python package changes replace all three of the old bdvcli, bdmacro and bd_vlib packages.  There are three modes of invoking the new package based on the usecase

1. As a CLI with built in help. 
```
$ configcli
Configuration CLI 1.0.

ccli> help

Documented commands (type help <topic>):
________________________________________
EOF  baseimg  ccli  exit  help  macro  namespace

ccli> help baseimg
usage: baseimg [-h] {version} ...

Baseimage related information.

optional arguments:
  -h, --help  show this help message and exit

Subcommands:
  {version}
    version   Returns the BlueData's base image version used to build the
              container image where this command is executed. Returns None if
              the container image is not based on the BlueData's base image.
ccli> help baseimg version
usage: baseimg version [-h]

optional arguments:
  -h, --help  show this help message and exit
ccli>
```
2. As a commandline utility from shell script or such.
```
$> ccli namespace node role_id
```
To maintain backward compatiblity and to minimize changes to existing app config scripts, I included most of the commandline arguments.
```
$ $ ccli --help
usage: ccli [-h] [-v] [--baseimg_version] [-g KEY] [-w KEY]
            [--tokenwait TOKEN] [--fqdn FQDN] [-t SECONDS] [--tokenwake TOKEN]
            [--success] [--error] [--service_key KEY] [--systemv SERVICE]
            [--systemctl SERVICE] [--unregister_srvc KEY] [--cp] [--node NODE]
            [--src SRC] [--dest DEST] [--perms PERMS] [--execute]
            [--remote_node REMOTE_NODE] [--script SCRIPT] [--tenant_info]
            [--tenant_info_lookup TENANT_INFO_KEY] [--get_local_group_fqdns]
            [--get_nodegroup_fqdns NODEGROUP_ID] [--get_all_fqdns]
            ...

Available options

positional arguments:
  instruction           A single instruction to execute.

optional arguments:
  -h, --help            show this help message and exit
  -v, --version         Prints configcli version and exits.

Backward compatible options:
  These command line options are provided to preserve backward
  compatibility.

  --baseimg_version     Returns the base image version used tobuild this
                        application image.
  -g KEY, --get KEY     Parse the namespace key delimited by a dot ('.') and
                        return its value.
  -w KEY, --wait KEY    A dot delimited namespace KEY that specifies services
                        to wait for. The KEY may either identify a specific
                        service or a group of services but not both at the
                        same time.
  --tokenwait TOKEN     A token to wait for on the specified node. Will
                        unblock when an explicit wake is called.
  --fqdn FQDN           The fully qualified domain name of the host on which
                        the corresponding wake is expected
  -t SECONDS, --timeout SECONDS
                        Maximum time (specified in seconds) to wait for a
                        response when -w/--wait API is invoked.
  --tokenwake TOKEN     Wake up all cluster processes waiting on the given
                        token
  --success             Indicates that waiters would be woken with success
                        status. This is the default if nothing is specified
  --error               Indicates that waiters would be woken with error
                        status
  --service_key KEY     A key to uniquely identify an application's service
                        being registered.
  --systemv SERVICE     A SystemV service name who's lifecycle is to be
                        managed by vAgent. The "service" command will be used
                        for handling the lifecycle events.
  --systemctl SERVICE   A SystemD service name who's lifecycle is to be
                        managed by vAgent. The 'systemctl' command will be
                        used for handling the lifecycle events.
  --unregister_srvc KEY
                        A namespace key uniquely identifying the application's
                        service to unregister.
  --cp                  Copy file to a node. File owner (user) & permissions
                        can be specified
  --node NODE           Destination node
  --src SRC             Absolute path to local file
  --dest DEST           Absolute path to destination file
  --perms PERMS         Permissions of the file in octal form.
  --execute             Execute a file on a remote node.
  --remote_node REMOTE_NODE
                        Node on which to execute the file.
  --script SCRIPT       Absolute path of the file to execute.
  --tenant_info         Get the tenant specific information as key-value pairs
  --tenant_info_lookup TENANT_INFO_KEY
                        Lookup the value of the given key in the
                        tenantinformation.
  --get_local_group_fqdns
                        Get all FQDNs deployed for the node group that the
                        local node belongs to.
  --get_nodegroup_fqdns NODEGROUP_ID
                        Get all FQDNs in the given Node group.
  --get_all_fqdns       Get FQDNs of all the nodes in the cluster.
```
3.  As a python library imported in some other module.
```
#
# setup.py
# 

from configcli import ConfigCli
ccli = ConfigCli(interactive=False)

namespace = ccli.getCommandObject('namespace')
distroId = namespace.getWithTokens(["node", "distro_id"])
...
...

macros = ccli.getCommandObject('macro')
nodeMacros = macros.getSubcommandObject('node')
selfIndex = nodeMacros.getNodeIndexSelf()
...
```

### ConfigCli Extensions
Some commands that need an `agent` are not implemented. Although, the corresponding cmdline arguments are included and will result in a failure if invoked. Those commands will be implemented as part of the `agent` if/when we do it. For now, the code related to extension importing is disabled.

## Fix for node index generation
For apps that configure services like zookeeper its useful to know the node's index during orchestration. However, its not always easy to compute a unique index that doesn't change when the cluster is expanded. 

To facilitate those apps' orchestration, we assign a monotonic id to each pod that is deployed in the cluster. And the node macros use this id to calculate an index. This way all existing nodes retain their current indexes during an expansion. Its important to point out that acluster shrink may not retain the index for all nodes depending on which node is being deleted. 

All the `go` code changes in this PR are to handle this feature.

## Testing. 
All existing apps are updated to use configCli and tested.

### CODE REVIEW

**Comment by joel-bluedata** [comment]
So far, I've looked through the go-code changes. Those are good.

The use of sync/atomic isn't necessary at the moment, but it does give us room to parallelize the role handling in syncRoles at some future time.

File: nodeprep/setup.py [hunk 1]
```diff
@@ -12,11 +14,41 @@
...
+    version = bdvcli.__version__,
+    description = '',
+
+    zip_safe=False,
+    include_package_data=True,
+
+    author = 'BlueData Software, Inc.',
+    author_email = 'support@bluedata.com',
+    url = 'https://github.com/bluek8s/kubedirector/nodeprep',
+    keywords = [ 'BlueData', 'vcli', 'bdmacro', 'EPIC', 'k8s', 'kubedirector'],
```
Comment (swamibluedata) [inline nodeprep/setup.py]: is this ok? References to EPIC and BlueData

File: pkg/apis/kubedirector.bluedata.io/v1alpha1/types.go [hunk 2]
```diff
@@ -100,6 +101,7 @@ type MemberStatus struct {
 	Service string `json:"service"`
 	PVC     string `json:"pvc,omitempty"`
 	State   string `json:"state"`
+	NodeId  string `json:"node_id"`
 }
```
Comment (joel-bluedata) [inline pkg/apis/kubedirector.bluedata.io/v1alpha1/types.go]: Given the changes to the status object (and subobjects), the schema in crd-cluster.yaml will need to be updated too. (And then the [wiki](https://github.com/bluek8s/kubedirector/wiki/Type-Definitions-for-KubeDirectorCluster), which is on me.)

**Comment by kcsraju** [comment]
Completed migrating `bdvcli` to `configcli`. Ready for a review now.

**Comment by joel-bluedata** [comment]
Yeah I think that is the only comment I have. I had looked over the go code changes before and that's fine. For the Python changes it was more of a "skim" on my part, but no red flags and the visible functionality changes are neat.

One heads-up/warning:

Tom K. [made a change recently](https://github.com/bluek8s/kubedirector/pull/96) to help flag any changes in the imported packages (and also to prepare us for eventually moving the KD codebase to using "go modules"). This means that, once you get this branch up to date with master, you may need to do "make modules" (and commit the resulting go.mod/go.sum) for Travis CLI to pass the build on this branch. In turn, that means you need to be using go version 11 so it can do modules stuff.

Swami is having [some issue with this process](https://github.com/bluek8s/kubedirector/pull/93) on his events branch so there might be a problem with that check that needs ironing out... check with Tom K if something about that is not working.

**Comment by kcsraju** [comment]
@joel-bluedata  I converted the `node_id` to int64 but, the change is pending testing. I'll drop a comment as soon as the testing is done.

**Comment by joel-bluedata** [comment]
I don't see any Travis CI activity at https://travis-ci.com/bluek8s/kubedirector/pull_requests . I think that perhaps it won't kick off a build for a non-mergable PR (and this one has conflicts).

**Comment by kcsraju** [comment]
Testing checks out too.

CDH Before expansion | After expansion
--------------------- | --------------------
screen shot 2018-10-25 at 12 37 02 pm | screen shot 2018-10-25 at 1 49 32 pm


Spark Before expansion | After expansion
--------------- | -----------------
screen shot 2018-10-25 at 3 52 52 pm | screen shot 2018-10-25 at 3 56 56 pm

**Pull request merged**